File: Sources/ccmux/Services/HookTrace.swift
import Foundation

/// Appends the native app's notification decisions to the shared hook trace at
/// `~/Library/Logs/ccmux-hooks.jsonl`.
///
/// The same file is written by `hooks/ccmux-notify.sh` (every hook Claude Code
/// fires) and by the daemon (`internal/hooktrace`), so one `tail -f` shows a hook
/// arriving, where it routed, and whether it became a notification. This type is
/// the app's third of that: it records the local macOS alert — posted, or
/// suppressed because you were already looking at the workspace.
///
/// Writing is best-effort in every direction. A trace that fails must never change
/// what the app shows, so there is no error to handle and nothing to report.
enum HookTrace {
    /// Overridable via `CCMUX_HOOK_TRACE`, the same variable the daemon and the
    /// hook script read, so pointing all three at one scratch file is one export.
    ///
    /// Read on each write rather than resolved once: an override that only takes
    /// effect if it was set before the first trace line is a trap, and tests need
    /// to redirect this away from the real log they'd otherwise pollute.
    static var path: String {
        if let override = ProcessInfo.processInfo.environment["CCMUX_HOOK_TRACE"], !override.isEmpty {
            return override
        }
        let home = FileManager.default.homeDirectoryForCurrentUser
        return home.appendingPathComponent("Library/Logs/ccmux-hooks.jsonl").path
    }

    /// Matches `internal/hooktrace.maxBytes`. Past it the file restarts: this is a
    /// tail buffer for a debugging session, not history worth keeping.
    private static let maxBytes: UInt64 = 8 << 20

    private static let queue = DispatchQueue(label: "com.ccmux.hooktrace")

    /// Local time with an offset, matching what the daemon (Go RFC 3339) and the
    /// hook script (Python `astimezone().isoformat()`) write. `ISO8601DateFormatter`
    /// defaults to GMT, which would put the app's lines an hour or more away from
    /// everyone else's in the same file — and timestamps are the only way to
    /// correlate a push line, which carries no trace id.
    private static let timestamp: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        f.timeZone = TimeZone.current
        return f
    }()

    /// Record one decision. `decision` is the one-word outcome the log is read
    /// for; `fields` carries whatever else that branch knows.
    static func write(decision: String, fields: [String: String] = [:]) {
        var line: [String: String] = [
            "ts": timestamp.string(from: Date()),
            "stage": "local",
            "decision": decision,
        ]
        line.merge(fields) { _, new in new }

        queue.async { append(line) }
    }

    private static func append(_ line: [String: String]) {
        guard let data = try? JSONSerialization.data(withJSONObject: line, options: [.sortedKeys]) else { return }
        var payload = data
        payload.append(0x0A) // newline

        let fd = open(path, O_WRONLY | O_APPEND | O_CREAT, 0o644)
        guard fd >= 0 else { return }
        defer { close(fd) }

        var st = stat()
        if fstat(fd, &st) == 0, UInt64(st.st_size) > maxBytes {
            _ = ftruncate(fd, 0)
        }
        // One write() per line. O_APPEND makes it atomic against the daemon and
        // the hook script appending to the same file at the same moment.
        _ = payload.withUnsafeBytes { buf in
            Darwin.write(fd, buf.baseAddress, buf.count)
        }
    }
}

File: daemon/internal/hooktrace/hooktrace.go
// Package hooktrace appends one JSON object per line to a shared trace file so a
// Claude Code hook can be followed from the moment it fires to the notification
// it did or didn't produce.
//
// Four processes write the same file: ccmux-notify.sh (every hook, before any
// filtering), the daemon's hook listener (what the event routed to), the daemon's
// push notifier (who got buzzed and who was suppressed), and the native app's
// listener (the local macOS alert). Each line carries a stage so the log reads as
// one story per hook rather than four disjoint logs.
//
// Correlation is by trace_id where the id survives the hop (script → socket →
// router) and by timestamp + workspace where it doesn't: the push notifier
// consumes the manager firehose, which carries a workspace and an attention state
// but no hook identity. That's a deliberate limit — threading an id through the
// whole event pipeline costs more than reading two adjacent lines.
package hooktrace

import (
	"encoding/json"
	"os"
	"path/filepath"
	"sync"
	"time"
)

// Stage names — the shared vocabulary of the trace file. Go writes two of them;
// the other two name what the shell script and the native app write, so the whole
// format is documented in one place even though this package never emits them.
const (
	StageHook  = "hook"  // ccmux-notify.sh: every hook Claude Code fired
	StageRoute = "route" // daemon hook listener: hook → pane + attention
	StagePush  = "push"  // daemon notifier: one line per subscription considered
	StageLocal = "local" // native app listener: the macOS alert it posted or suppressed
)

// maxBytes caps the trace file. Past it the file is truncated and started fresh
// rather than rotated: this is a debugging aid you tail, not an audit log worth
// keeping history for.
//
// Truncation can't corrupt the file — an O_APPEND write always seeks to the
// current end, so nobody writes into a hole — but with four processes appending
// it is not only old lines that go. Whatever another writer appended between this
// process's size check and its truncate is discarded too. Losing a few lines at
// the 8 MB mark is the accepted cost of not rotating.
const maxBytes = 8 << 20

// Line is one traced event. Everything is omitempty because each stage fills a
// different subset; a line is meant to be read, not schema-validated.
type Line struct {
	TS      string `json:"ts"`
	Stage   string `json:"stage"`
	TraceID string `json:"trace_id,omitempty"`
	Event   string `json:"event,omitempty"`

	// Routing (StageRoute).
	CWD       string `json:"cwd,omitempty"`
	SessionID string `json:"session_id,omitempty"`
	PaneID    string `json:"pane_id,omitempty"`
	Resolved  string `json:"resolved_pane,omitempty"`
	Attention string `json:"attention,omitempty"`
	Session   string `json:"session_signal,omitempty"`

	// Push (StagePush).
	WorkspaceID string `json:"workspace_id,omitempty"`
	Login       string `json:"login,omitempty"`
	Suppressed  string `json:"suppressed_by,omitempty"`

	// Outcome, in every stage: what actually happened, in one word.
	Decision string `json:"decision"`
	Detail   string `json:"detail,omitempty"`
}

var (
	mu   sync.Mutex
	path = defaultPath()
)

// DefaultPath is where the trace lands. It sits beside the daemon's own log so
// one directory holds everything you'd tail while debugging notifications.
func DefaultPath() string {
	mu.Lock()
	defer mu.Unlock()
	return path
}

// SetPath redirects the trace, for tests.
func SetPath(p string) {
	mu.Lock()
	defer mu.Unlock()
	path = p
}

func defaultPath() string {
	if p := os.Getenv("CCMUX_HOOK_TRACE"); p != "" {
		return p
	}
	home, err := os.UserHomeDir()
	if err != nil {
		return filepath.Join(os.TempDir(), "ccmux-hooks.jsonl")
	}
	return filepath.Join(home, "Library", "Logs", "ccmux-hooks.jsonl")
}

// Write appends one line. Every failure is swallowed: a trace that can't be
// written must never change what the daemon does with a hook.
func Write(l Line) {
	l.TS = time.Now().Format(time.RFC3339Nano)
	data, err := json.Marshal(l)
	if err != nil {
		return
	}
	data = append(data, '\n')

	mu.Lock()
	defer mu.Unlock()
	f, err := os.OpenFile(path, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0o644)
	if err != nil {
		return
	}
	defer f.Close()
	if st, err := f.Stat(); err == nil && st.Size() > maxBytes {
		_ = f.Truncate(0)
	}
	_, _ = f.Write(data) // one write per line: interleaves atomically with the shell's
}
